### PR TITLE
limit `change_visibility` assist to applicable items

### PR DESCRIPTION
**Pull request by 1Kinoti**
this pr limits the `change_visibility` assist to applicable items. top level items in this context means items that are not nested within `fn`s or `trait`s.

now
```rs
fn foo {
    // assists on this `struct` keyword won't include `change_visibility`
    struct Bar {} 
}

trait Foo {
   // same with the `fn` here
   fn bar();
}
```

### CODE REVIEW

File: crates/ide-assists/src/handlers/change_visibility.rs [hunk 1]
```diff
@@ -46,16 +47,36 @@ fn add_vis(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
 
     let (offset, target) = if let Some(keyword) = item_keyword {
         let parent = keyword.parent()?;
-        let def_kws =
-            vec![CONST, STATIC, TYPE_ALIAS, FN, MODULE, STRUCT, ENUM, TRAIT, USE, MACRO_DEF];
-        // Parent is not a definition, can't add visibility
-        if !def_kws.iter().any(|&def_kw| def_kw == parent.kind()) {
+
+        let can_add = |node: &SyntaxNode| -> bool {
```
Comment (lowr) [inline crates/ide-assists/src/handlers/change_visibility.rs:51]: `can_add` doesn't need to be a closure I'd say. Can you make this a proper function or inline the body even as it's only used once?

File: crates/ide-assists/src/handlers/change_visibility.rs [hunk 2]
```diff
@@ -46,16 +47,36 @@ fn add_vis(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
...
+            const LEGAL: &[syntax::SyntaxKind] =
```
Comment (lowr) [inline crates/ide-assists/src/handlers/change_visibility.rs:52]: Const-ifying is nice :+1: Let's import `SyntaxKind`.

File: crates/ide-assists/src/handlers/change_visibility.rs [hunk 3]
```diff
@@ -46,16 +47,36 @@ fn add_vis(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
...
+                let is_legal = |k| matches!(k, SOURCE_FILE | MODULE);
+                let Some(p) = node.parent() else {
+                    return false;
+                };
+
+                if p.kind() == ASSOC_ITEM_LIST {
+                    p.parent()
+                        .filter(|imp| imp.kind() == IMPL)
+                        .and_then(|imp| imp.parent())
+                        .is_some_and(|it| is_legal(it.kind()))
```
Comment (lowr) [inline crates/ide-assists/src/handlers/change_visibility.rs:65]: Can you check if trait impls (i.e. `impl Trait for Type`) don't bypass this check? They are also parsed as `IMPL` and we don't want to provide this assist for them I suppose.

**Review by lowr — COMMENTED** [review]
Yeah, I agree that we don't want to provide this assist for nested items. Left several suggestions.

File: crates/ide-assists/src/handlers/change_visibility.rs (hunk 3 above)
Comment (lowr) [reply crates/ide-assists/src/handlers/change_visibility.rs:65]: Ah, can you actually exempt inherent (non-trait) impls from this check and make them "legal" even when they're not at top-level? Impls have non-local effect and it's reasonable to provide the assist for cases like this:

```rust
const _: () = {
    impl S {
        fn/*$0*/ foo() {} 
    }
};
```

**Comment by lowr** [comment]
I'd like to see two more things, but otherwise looks good to me!

- Can you add a comment in `fn can_add()` explaining why inherent impls are exempted?
- We generally don't want to include merge commits. Can you remove a81498a3d48b0f04346bf0d28f8717fd4a8cce61 and maybe squash all the commits into one?

**Comment by 1Kinoti** [comment]
@lowr done

**Comment by lowr** [comment]
Nice, thanks!

@bors r+

**Pull request merged**